### PR TITLE
feat(cmf): add script to merge settings

### PR DESCRIPTION
**Pull request by romainseb**
**What is the problem this PR is trying to solve?**
Currently, we have to define some script in each project to merge settings when needed
**What is the chosen solution to this problem?**
Add script in "bin" of cmf. This allow us to launch "cmf-settings" in a teminal when cmf is installed on a project. With some cmf.json file to load the definition, we can have one script at one place to merge settings from differents projects / ressources

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/cmf/scripts/cmf-settings.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// 2 - Get sources & destination paths
+const sources = program.dev
+	? cmfconfig['cmf-settings']['dev-sources']
+	: cmfconfig['cmf-settings'].sources;
+const destination = `${process.cwd()}/${cmfconfig['cmf-settings'].destination}`;
+
+// 3 - Extract json from sources
+log('\nExtracting configuration from : \n');
+sources.forEach(source => {
+	findJsonInFolder(`${process.cwd()}/${source}`).forEach(filePath => {
```
Comment (jare-talend) [inline packages/cmf/scripts/cmf-settings.js]: use the librairies path to manage the / on windows, linux

File: packages/cmf/scripts/cmf-settings.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const overrideActions = utils.overrideActions;
+const log = utils.getLogger(program.quiet);
+const configurations = [];
+const cmfconfig = utils.getCmfConfig();
+
+// 2 - Get sources & destination paths
+const sources = program.dev
+	? cmfconfig['cmf-settings']['dev-sources']
+	: cmfconfig['cmf-settings'].sources;
+const destination = `${process.cwd()}/${cmfconfig['cmf-settings'].destination}`;
```
Comment (jare-talend) [inline packages/cmf/scripts/cmf-settings.js]: use the librairies path to manage the / on windows, linux
https://nodejs.org/api/path.html

File: packages/cmf/scripts/utils/index.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	};
+}
+
+/**
+ * This function scan the current folder to search for cmf.json file
+ * & return the content of this file
+ */
+function getCmfConfig() {
+	try {
+		return require(`${process.cwd()}/cmf.json`); // eslint-disable-line
```
Comment (jare-talend) [inline packages/cmf/scripts/utils/index.js]: use path to manage / on  windows, linux

File: packages/cmf/scripts/utils/index.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	return currentChild;
+}
+
+/**
+ * This function allow to override some route settings
+ * @param {string} path The path of the route to override
+ * @param {object} settings the settings of a cmf application
+ */
+function overrideRoutes(path, settings) {
+	const childRoutes = getChildRoutes(path.split('/'), settings);
```
Comment (jare-talend) [inline packages/cmf/scripts/utils/index.js]: manage / on windows, linux with https://nodejs.org/api/path.html#path_path_sep

File: packages/cmf/README.md [hunk 5]
```diff
@@ -168,6 +168,32 @@ Note onResponse and onError accept function:
...
 
+## Scripts
+
+When you have cmf in you package.json, you can run in your project scope this script :
+- cmf-settings
+
+It require a cmf.json file with this format in your webapp's project root :
+```json
+{
+	"cmf-settings": {
```
Comment (jare-talend) [inline packages/cmf/README.md]: could you beautify the sample

File: packages/cmf/scripts/cmf-settings.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+// 3 - Extract json from sources
+log('\nExtracting configuration from : \n');
+sources.forEach(source => {
```
Comment (jare-talend) [inline packages/cmf/scripts/cmf-settings.js]: use map instead forEach.

File: packages/cmf/scripts/utils/index.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	settings.overrideRoutes[path].forEach(config => {
```
Comment (jare-talend) [inline packages/cmf/scripts/utils/index.js]: use map instead forEach

File: packages/cmf/scripts/utils/index.js (hunk 4 above)
Comment (romainseb) [reply packages/cmf/scripts/utils/index.js]: Here, we have urls, not file system path ;)

File: packages/cmf/scripts/cmf-settings.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	Object.keys(settings.overrideActions).forEach(id => {
+		overrideActions(id, settings);
+	});
+}
+
+// 6 - Write the merged file
+log(`Merge to ${destination}`);
+mkdirp(pathLib.dirname(destination), () => {
+	fs.writeFile(destination, JSON.stringify(settings), () => {
+		console.log('Merged');
```
Comment (ncomont) [inline packages/cmf/scripts/cmf-settings.js]: Why didn't you use your `logger` ?

Comment (romainseb) [reply packages/cmf/scripts/cmf-settings.js]: oh i forgot this one, thanks

File: packages/cmf/README.md [hunk 9]
```diff
@@ -168,6 +168,33 @@ Note onResponse and onError accept function:
...
+## Scripts
+
+When you have cmf in you package.json, you can run in your project scope this script :
+- cmf-settings
+
+It require a cmf.json file with this format in your webapp's project root :
+
+```json
+{
+	"cmf-settings": {
```
Comment (jmfrancois) [inline packages/cmf/README.md]: Because the file is named cmf.json I think settings is enough

File: packages/cmf/README.md [hunk 10]
```diff
@@ -168,6 +168,33 @@ Note onResponse and onError accept function:
...
+		"dev-sources": [
+			"src/settings",
+			"../../dataset/webapp/src/settings"
+		],
+		"sources": [
+			"src/settings",
+			"node_modules/@talend/dataset/lib/settings"
+		],
+		"destination": "src/assets/cmf-settings.json"
```
Comment (jmfrancois) [inline packages/cmf/README.md]: I would go with
* "source"
* "source-dev"

File: packages/cmf/package.json [hunk 11]
```diff
@@ -27,8 +27,12 @@
     "type": "git",
     "url": "https://github.com/Talend/ui.git"
   },
+  "bin": {
+    "cmf-settings": "./scripts/cmf-settings.js"
+  },
   "dependencies": {
     "babel-polyfill": "6.26.0",
+    "deepmerge": "^1.5.1",
```
Comment (jmfrancois) [inline packages/cmf/package.json]: please add this to versions.js

File: packages/cmf/scripts/utils/index.js (hunk 7 above)
Comment (jare-talend) [reply packages/cmf/scripts/utils/index.js]: childRoutes.concat(settings.overrideRoutes[path]) instead

**Pull request merged**